resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryImpl.java

```diff
@@ -139,12 +139,18 @@ private ContextImpl() {
         @Override
 		public void onSuccess() {
 			int currentNumOfAttempts = numOfAttempts.get();
-			if (currentNumOfAttempts > 0) {
+			if (currentNumOfAttempts > 0 && currentNumOfAttempts < maxAttempts) {
 				succeededAfterRetryCounter.increment();
 				Throwable throwable = Option.of(lastException.get()).getOrElse(lastRuntimeException.get());
 				publishRetryEvent(() -> new RetryOnSuccessEvent(getName(), currentNumOfAttempts, throwable));
 			} else {
-				succeededWithoutRetryCounter.increment();
+				if (currentNumOfAttempts >= maxAttempts) {
+					failedAfterRetryCounter.increment();
```

Comment:
Looks a bit weird that there is a branch inside onSuccess method which emits error event. It would be more consistent to rename onSuccess to onOperationFinished or onRetriesStopped etc.

Reply:
Yes, we are also discussing it internally.
Unfortunately we don't want to introduce a non-backward compatible change. The method `onSuccess` is part of the public API.

Reply:
The onResult handling was added at a later point in time. That's why it doesn't really fit into the public API.

Reply:
So what is the final conclusion here :) ?

Reply:
We could also just create a new `onResult` and move the logic, which is currently in the decorators, into that method.
The code is actually duplicated everywhere:

```
T result = supplier.get();
				final boolean validationOfResult = context.onResult(result);
				if (!validationOfResult) {
					context.onSuccess();
					return result;
				}
```